Add MFEM LOR options to the multires control operator

JustinPrivitera opened the pull request:
### Description

Resolves #20634 <!-- If this PR is unrelated to a ticket, please erase this line -->
Resolves #19046
Resolves #19469
Resolves #19121

**Reviewers: please have a look at the changes to avtGenericDatabase and the File Format Interface classes. Please also see my new documentation, linked below.**

<!-- Please include a summary of the change and any other information and instructions to help reviewers understand the work -->
- Enhance Blueprint plugin, MFEM plugin, MFEM data adaptor, Multires control operator to allow for customizable visualization of high-order data.
- New options have been added for mesh refinement, field projection, and basis choice.
- I encourage reviewers to read my new [documentation](https://visit-sphinx-github-user-manual--20666.org.readthedocs.build/en/20666/using_visit/Operators/OperatorTypes/MultiresControl_operator.html) to understand what has changed.
- I also made significant changes to avtGenericDatabase and the file format interface classes to allow the MFEM refinement options to change variable centering. Currently centering changes are only enabled for Blueprint and MFEM, but changes had to be made at a deep level to enable this.

<!-- Note that all the checklist items in various bulleted lists below end with ~~ characters.
     This is to facilitate striking them out when they do not apply.
     One just has to add ~~ characters just before the opening square bracket [ -->

### Type of change

<!-- Please check one of the boxes below -->

* [x] Bug fix~~
* [x] New feature~~
* [x] Documentation update~~
* [x] Other~~ <!-- please explain with a note below -->
infrastructure change

### How Has This Been Tested?

<!-- Please describe the tests you've added or any tests that already cover this change. Include relevant information, such as which operating system you tested on. -->
new tests; built and ran on rzhound toss4; all tests new and old pass

### Reminders:

- Please follow the [style guidelines][1] of this project.
- Please perform a self-review of your code before submitting a PR and asking others to review it.
- Please assign reviewers (see [VisIt's PR procedures][2] for more information).

### Checklist:

<!-- For items in this checklist that do not apply, simply insert two tilde chars, `~~`, just ahead of the left bracket char, `[` at the beginning of a line. Each line ends with two tilde chars to make doing such ~~strikeouts~~ easy. -->

- [x] I have commented my code where applicable.~~
- [x] I have updated the release notes.~~
- [x] I have made corresponding changes to the documentation.~~
- [x] I have added debugging support to my changes.~~
- [x] I have added tests that prove my fix is effective or that my feature works.~~
- [x] I have confirmed new and existing unit tests pass locally with my changes.~~
- [x] I have added new baselines for any new tests to the repo.~~
- ~~[ ] I have NOT made any changes to [*protocol* or *public interfaces*][3] in an RC branch.~~

[1]: https://visit-sphinx-github-user-manual.readthedocs.io/en/develop/dev_manual/StyleGuide.html
[2]: https://visit-sphinx-github-user-manual.readthedocs.io/en/develop/dev_manual/pr_create.html#reviewers
[3]: https://visit-sphinx-github-user-manual.readthedocs.io/en/develop/dev_manual/RCDevelopment.html#communication-protocols-and-public-apis

JustinPrivitera committed:
add string

JustinPrivitera committed:
plumb refinement method through

JustinPrivitera committed:
rerun xml gen tools

JustinPrivitera committed:
hook a few more things up

JustinPrivitera committed:
now blueprint uses operator refinement instead of db option refinement

JustinPrivitera committed:
leave refinement method as an open option

JustinPrivitera committed:
mfem enabled as well

JustinPrivitera commented:
@cyrush how much does this conflict with your ongoing MFEM work?

I will hold off on adding modification comments and finishing this until we can discuss if there are structural changes needed, especially with regard to your current MFEM work.

JustinPrivitera committed:
update tests and name of open setting

JustinPrivitera committed:
revert the unnecessary name change for backwards compatibility

JustinPrivitera committed:
fixed the tests

JustinPrivitera commented:
The way I have set up the refinement method choices is not ideal. I should change it to not use strings.

JustinPrivitera commented:
we need to have likely 4 options: legacy, lor, lor force nodal, lor force zonal

cyrush commented:
Possible names for the cases:
```
LOR Projection (Default)
Discontinuous Refine
LOR Nodal Projection
LOR Zonal Projection
```

JustinPrivitera committed:
support the 4 choices

JustinPrivitera committed:
use new options

JustinPrivitera committed:
fix mfem tests

JustinPrivitera commented:
TODO
- [x] We want to put the enum into the mfem data adaptor and have the switchyard for the 4 cases take place where the current switchyard is for legacy and new refine.
- [x] We also want to support backwards compatibility with the default file open options.

JustinPrivitera committed:
pipe the LOR options further down

JustinPrivitera committed:
Merge branch 'develop' into task/JustinPrivitera/10_15_25/multires_operator_refine_method

JustinPrivitera committed:
clean up mfem tests

JustinPrivitera commented:
@cyrush I've moved this forward; I have added a switchyard down in the MFEM data adaptor for the 4 cases. 2 of the cases map onto our previously existing cases; the others currently produce errors. I imagine that we will need the MFEM folks to tell us how to implement some of the cases?

JustinPrivitera committed:
fix tests

IdoAkkerman commented:
I think line 1033 has a small bug. I think it should be:

` mfem::FiniteElementSpace lo_fes(&lo_mesh, lo_col, gf->FESpace()->GetVectorDim());`

This should fix #18700


The difference between `gf->FESpace()->GetVectorDim()` and `gf->FESpace()->GetVDim()` is that the first will give the
number of spatial dimensions spanned by all the finite element, while the second provides the number of finite elements defined on the mesh. So incase of a vector element (Hdiv/RT or Hcurl/ND) these number will not be the same!!

IdoAkkerman commented:
In MFEM I just made a PR that would allow `GetTransferOperator` to work with more generic cases (only tested for NURBS).

JustinPrivitera commented:
@IdoAkkerman please see continued discussion in #19046

JustinPrivitera committed:
add TODOs

JustinPrivitera committed:
matrix free version of the transfer operator

JustinPrivitera committed:
comment

JustinPrivitera committed:
fix zonal nodal typo

JustinPrivitera committed:
added vector to l2

JustinPrivitera committed:
skeleton for correct path later

JustinPrivitera committed:
merge in develop

JustinPrivitera committed:
refinement methods implemented

JustinPrivitera committed:
remove todos

JustinPrivitera committed:
change to closed uniform basis type where possible

IdoAkkerman commented on src/avt/MFEM/avtMFEMDataAdaptor.C:
```diff
@@ -989,50 +1016,70 @@ avtMFEMDataAdaptor::RefineGridFunctionToVTK(mfem::Mesh *mesh,
+
+    mfem::GridFunction *gf_to_use = gf;
+    bool delete_gf_to_use = false;
+    if (hdiv || hcurl)
```
I like the idea of splitting the problem in smaller subproblems. So mapping the VectorFE on an L2 space is great in achieving this.

However, in case of NURBS-based VectorFE space all solution components are at least C0 continous (potentially higher). So mapping on L2 seems a bit wasteful(more internal mem usage and computational cost) and might even incorrectly give a discontinuous field.

I think it would be appropriate to add a check for NURBS

`bool nurbs = basis.find("NURBS_") != std::string::npos;`

and map the NURBS based VectorFE on a H1 space using 

```
mfem::GridFunction *
ConvertVectorFEToH1(mfem::GridFunction *vec_gf)
{
  ....
}

```

There are some subtleties  to keep track of. I would not mind providing the above suggested changes however I can not get the git version of visit compiled because of some improper QT6 configuration stuff. I do not know it my mint 21 machine is incorrectly configure or whether the build scripts are incorrect....
If somebody could assist me with this I could provide the suggested code.

Should I open an issue related to the building issues?

IdoAkkerman replied:
Fingers crossed but I think I got it to work...

IdoAkkerman replied:
I got it working it compiled, and was able to make changes to get the NURBS scalar case to work. The Hdiv NURBS case did not work, because the Magnitude Expression thinks we are dealing with a Scalar.

JustinPrivitera replied:
So I guess I am confused by what you're suggesting. After a lot of discussion with MFEM team members and lots of our own thinking, we'd like to let users choose from 4 options for how LOR is performed. They can ask for discontinuous refinement, which seems to be battle-tested for all the different cases and is what we have used for years. They can also ask for nodal and zonal projection using the newer refinement method. Forcing zonal or nodal results requires transforming the grid functions appropriately. They can also choose a default, which will try to pick the best result depending on what they are trying to visualize.

To let users force a zonal or nodal projection, we have to transform grid functions to low order H1 (nodal) or L2 (zonal) variables. The path suggested to us by other MFEM team members was to convert Hdiv and Hcurl grid functions to L2, and then convert from L2 to H1 if needed. It sounds like you are suggesting that there are cases for which this is not ideal, and that there may be machinery in place to convert directly from a vector grid function to H1?

So just to be clear, you want us to check for `NURBS_` in the basis name and use `ConvertVectorFEToH1` only in the case that nodal (H1) results were requested. Is that correct? Are there other cases for which we should use `ConvertVectorFEToH1`? If we could skip converting intermediately to L2 that would be best, right?

IdoAkkerman replied:
Sorry for the late reply. But I think the -- already accepted PR -- looks fine.
The NURBS  solution are always at least C0 continuous, so no need to project on L2.

IdoAkkerman commented on src/avt/MFEM/avtMFEMDataAdaptor.C:
```diff
@@ -989,50 +1016,70 @@ avtMFEMDataAdaptor::RefineGridFunctionToVTK(mfem::Mesh *mesh,
     bool h1 = basis.find("H1_") != std::string::npos;
-    bool node_centered;
-    if (h1 && l2)
+    // TODO is this correct? We need test data
```
I can provide some data. What would be the best way to provide the data?

PS: I do not have push permission to the visit repo. Perhaps that is the easiest way. But I do not know your policies regarding this offcourse.

cyrush replied:
@IdoAkkerman Example files would really help, especially for the nurbs case. If the files are small enough you could attach them in zip files here. 

Earlier today, I created some hcurl and hdiv examples from ex3 and ex4 that we are going to use to test these changes shortly

JustinPrivitera commented:
Note for @cyrush and @JustinPrivitera: investigate the interplay with materials

IdoAkkerman commented:
I think I was able to fix it. 
The case for 1 Scalar NURBS and 1 HDiv NURBS worked.
These are the changes I made 
[avtMFEMDataAdaptor.cpp.patch](https://github.com/user-attachments/files/23660710/avtMFEMDataAdaptor.cpp.patch)

I will:
- Do more extensive tests
- Send the test data to you folks -- if interested.

JustinPrivitera committed:
honor user choices with periodic meshes

JustinPrivitera commented:
> I think I was able to fix it. The case for 1 Scalar NURBS and 1 HDiv NURBS worked. These are the changes I made [avtMFEMDataAdaptor.cpp.patch](https://github.com/user-attachments/files/23660710/avtMFEMDataAdaptor.cpp.patch)
> 
> I will:
> 
>     * Do more extensive tests
> 
>     * Send the test data to you folks -- if interested.

The patch fails; perhaps it is truncated? I will work to manually make the included changes.

Test data would be much appreciated!

JustinPrivitera committed:
Merge branch 'develop' into task/JustinPrivitera/10_15_25/multires_operator_refine_method

JustinPrivitera committed:
test data and correct basis string

JustinPrivitera commented:
@IdoAkkerman I have added your changes here: #20737 

All looks good to me but I will wait to merge into this branch until others have had a chance to look.

JustinPrivitera committed:
having trouble getting the old tests to pass

JustinPrivitera committed:
respect association guess

JustinPrivitera committed:
MFEM Patch from @IdoAkkerman

JustinPrivitera committed:
Merge branch 'develop' into task/JustinPrivitera/10_15_25/multires_operator_refine_method

JustinPrivitera commented:
Very slight differences:
<img width="919" height="921" alt="image" src="https://github.com/user-attachments/assets/5f7ec6fe-bdd6-414c-9012-5092b987f145" />

I don't think we should rebaseline, as I think these datasets should be taking the same path as before.

JustinPrivitera commented:
<img width="565" height="1026" alt="image" src="https://github.com/user-attachments/assets/2454e130-d92f-4f46-ae08-c361cc90592f" />

JustinPrivitera committed:
restore choice of override

JustinPrivitera committed:
revert more changes; the mystery is still unsolved

JustinPrivitera committed:
I conclude that we need to rebaseline

JustinPrivitera committed:
restore tests

IdoAkkerman commented:
@JustinPrivitera In the old code on line 1027 of avtMFEMDataAdaptor.C the refinement is made using `mfem::BasisType::GaussLobatto`, while in the new code on line 1148 it uses `mfem::BasisType::ClosedUniform`.

This will change the location of refinement points slightly, which might explain the practically invisible changes.
My suggestion would be:
1) Change the refinement back, ans see if the 'errors' disappear. 
2) Make an 'executive decision' which refinement is the correct one and change the reference files accordingly (if required)

JustinPrivitera committed:
basis type choice

JustinPrivitera committed:
thinking about which things to test

JustinPrivitera commented:
> @JustinPrivitera In the old code on line 1027 of avtMFEMDataAdaptor.C the refinement is made using `mfem::BasisType::GaussLobatto`, while in the new code on line 1148 it uses `mfem::BasisType::ClosedUniform`.
> 
> This will change the location of refinement points slightly, which might explain the practically invisible changes. My suggestion would be:
> 
>     1. Change the refinement back, ans see if the 'errors' disappear.
> 
>     2. Make an 'executive decision' which refinement is the correct one and change the reference files accordingly (if required)

Yes, we realized this. It turns out it that using a matrix free transfer operator also leads to tiny changes. I have added an option to choose the basis type now in the multires control operator. I am likely going to blow away most if not all of the existing LOR tests and replace them with new ones that test the full range of options now.

JustinPrivitera committed:
learnings from cyrus

JustinPrivitera committed:
separated field projection and mesh refine in most places

JustinPrivitera committed:
narrowed down multires options further

JustinPrivitera committed:
default mesh refinement method

JustinPrivitera committed:
changes to mfem data adaptor to use new enums

JustinPrivitera committed:
works, now just need to get it to compile

JustinPrivitera committed:
compiles, now just need to get it to work

JustinPrivitera committed:
vdim versus vectordim

JustinPrivitera committed:
legacy -> discontinuous

JustinPrivitera committed:
comments

JustinPrivitera committed:
mostly wired up centering change getvar version

JustinPrivitera committed:
support centering change in more places

JustinPrivitera committed:
scalar var forced recenter works now for mfem

JustinPrivitera committed:
blueprint centering change is wired up

JustinPrivitera committed:
getvectorvar can now adjust centering

JustinPrivitera committed:
symmetric tensors

JustinPrivitera committed:
carefully added centering change for all var types

JustinPrivitera committed:
secondary variables centering change supported

JustinPrivitera committed:
centering change for query

JustinPrivitera committed:
whitespace

JustinPrivitera committed:
remove outdated comment

JustinPrivitera committed:
gui text changes

JustinPrivitera committed:
gui is back to a good place

JustinPrivitera committed:
window tweak

JustinPrivitera committed:
begin docs

JustinPrivitera committed:
mfem plugin detect association

JustinPrivitera committed:
remove comment

JustinPrivitera committed:
move notes around

JustinPrivitera committed:
fix gamebreaking bug

JustinPrivitera committed:
Merge branch 'develop' into task/JustinPrivitera/10_15_25/multires_operator_refine_method

JustinPrivitera committed:
change prints to info

JustinPrivitera committed:
docs

JustinPrivitera committed:
docs fixes

JustinPrivitera committed:
Merge branch 'develop' into task/JustinPrivitera/10_15_25/multires_operator_refine_method

JustinPrivitera committed:
new baselines

JustinPrivitera committed:
mfem tests in progress

JustinPrivitera committed:
baselines

JustinPrivitera committed:
grid function tests

JustinPrivitera committed:
quadrature basis tests

JustinPrivitera committed:
hcurl hdiv tests and baselines

JustinPrivitera committed:
blueprint baselines

JustinPrivitera committed:
update blueprint baselines

JustinPrivitera committed:
change vector plot settings

JustinPrivitera committed:
new vector gf baselines

JustinPrivitera committed:
release notes

JustinPrivitera committed:
closeup shots

JustinPrivitera committed:
closeup tests

JustinPrivitera committed:
modification comments

JustinPrivitera committed:
Merge branch 'develop' into task/JustinPrivitera/10_15_25/multires_operator_refine_method

biagas commented on src/avt/Database/Formats/avtSTMDFileFormatInterface.C:
```diff
@@ -146,9 +146,30 @@ avtSTMDFileFormatInterface::GetMesh(int ts, int dom, const char *mesh)
+    }
+    else
+    {
+        return GetVar(ts, dom, var);
```
just curious why you call `GetVar` instead of `timesteps[ts]->GetVar` ?
Seems unnecessary since the logic from GetVar for checking `ts` has already been performed.
Similar question for GetVectorVar.

JustinPrivitera replied:
My mistake; thanks for catching!

biagas commented on src/databases/Blueprint/avtBlueprintOptions.C:
```diff
@@ -29,18 +29,15 @@
 //      Removed inaccurate comment.
 //
+//      Justin Privitera, Wed Dec 17 14:01:55 PST 2025
+//      Removed LOR options.
```
How about removing the comments from the bottom of the method?

JustinPrivitera replied:
Those comments predate my work here. I am inclined to leave them as an example for the future. They show up in the following plugins:
```
Blueprint
Cube
Geqdsk
GGCM
MFIXCDF
Miranda
MOAB
Nek5000
paraDIS
PLY
ProteinDataBank
STL
unv
Xmdv
```

But I believe they also appear when you regenerate the options file with the xml tools. I think it is useful to leave them around in case new options are added. But I could be convinced otherwise.

biagas replied:
No worries either way.

biagas commented on src/databases/MFEM/avtMFEMOptions.C:
```diff
@@ -21,19 +21,17 @@
 //  Creation:   Wed Oct 12 14:46:17 PST 2022
+// 
+//  Modifications:
+//    Justin Privitera, Wed Dec 17 14:01:55 PST 2025
```
Since there are no longer any read or write options, should this file be removed and references to Options in the other MFEM plugin files removed as well ?

JustinPrivitera replied:
Yes, removed. Thanks!

JustinPrivitera committed:
remove mfem options

cyrush reviewed (dismissed):
Looks good.
Thank you for these comprehensive changes and extensive tests.

JustinPrivitera committed:
change due to feedback

JustinPrivitera committed:
change due to feedback

JustinPrivitera committed:
Merge branch 'develop' into task/JustinPrivitera/10_15_25/multires_operator_refine_method

biagas reviewed (approved):
Great job Justin!

cyrush commented:
Awesome!
